### PR TITLE
Leading/Trailing throttle and throttleTime configuration

### PR DESCRIPTION
**Pull request by benlesh**
related to #1625  see commits for details.

### CODE REVIEW

File: spec/operators/throttleTime-spec.ts [hunk 1]
```diff
@@ -141,4 +142,32 @@ describe('Observable.prototype.throttleTime', () => {
...
+
+      const result = e1.throttleTime(t, rxTestScheduler, { leading: true, trailing: true });
+
+      expectObservable(result).toBe(expected);
+      expectSubscriptions(e1.subscriptions).toBe(e1subs);
+    });
+  });
+
+  describe('throttleTime(fn, { leading: false, trailing: true })', () => {
+    asDiagram('throttleTime(fn, { leading: true, trailing: true })')('should immediately emit the first value in each time window', () =>  {
```
Comment (david-driscoll) [inline spec/operators/throttleTime-spec.ts]: `{ leading: true, trailing: true }` should be `{ leading: false, trailing: true }`

File: spec/operators/throttle-spec.ts [hunk 2]
```diff
@@ -337,4 +337,40 @@ describe('Observable.prototype.throttle', () =>  {
...
+      const result = e1.throttle(() =>  e2, { leading: true, trailing: true });
+
+      expectObservable(result).toBe(expected);
+      expectSubscriptions(e1.subscriptions).toBe(e1subs);
+      expectSubscriptions(e2.subscriptions).toBe(e2subs);
+    });
+  });
+
+  describe('throttle(fn, { leading: false, trailing: true })', () => {
+    asDiagram('throttle(fn, { leading: true, trailing: true })')('should immediately emit the first value in each time window', () =>  {
```
Comment (david-driscoll) [inline spec/operators/throttle-spec.ts]: `{ leading: true, trailing: true }` should be `{ leading: false, trailing: true }`

**Comment by david-driscoll** [comment]
Is it fair to say...

* `x.throttleTime(100, { trailing: true })` is the same as `x.auditTime(100)`?
* `x.throttleTime(100, { leading: true, trailing: true })` is the same as `Observable.merge(x.throttleTime(100), x.auditTime(100))`?

Also would it be possible (perhaps as a separate PR!) to give the same treatment to `debounce` and `debounceTime`?

We have no operator that compares to [lodash's `debounce(x => {}, { leading: true})`](https://lodash.com/docs/4.17.4#debounce)

**Comment by benlesh** [comment]
> Also would it be possible (perhaps as a separate PR!) to give the same treatment to debounce and debounceTime?

That's the idea.

**Comment by benlesh** [comment]
@david-driscoll the changes are made. Good catch. I put them in a separate commit so I didn't have to pull everything apart and change two commits.

**Comment by david-driscoll** [comment]
Looks good to me.  @kwonoj @trxcllnt @jayphelps any complaints?

**Comment by benlesh** [comment]
The potential bug found in `throttle` (it existed prior, and this PR does NOT address it) I mention in my commit messages is listed here: #2466

**Comment by connorwyatt** [comment]
Hi, I would be interested in knowing when you guys think this will be in a release? We have a workaround that we can remove once this goes in and I was wondering when you think this will be merged.

Thanks

**Comment by lee-tunnicliffe** [comment]
I could also do with this change for my project. Any update would be greatly appreciated.

**Pull request merged**